gateway/src/apicast/policy/caching/policy.lua

```diff
@@ -0,0 +1,94 @@
+--- Caching policy
+-- Configures a cache for the authentication calls against the 3scale backend.
+-- The 3scale backend can authorize (status code = 200) and deny (status code =
+-- 4xx) calls. When it fails, it returns a 5xx code.
+-- This policy support three kinds of caching:
+--   - Strict: it only caches authorized calls. Denied and failed calls
+--     invalidate the cache entry.
+--   - Resilient: caches authorized and denied calls. Failed calls do not
+--     invalidate the cache. This allows us to authorize and deny calls
+--     according to the result of the last request made even when backend is
+--     down.
+--   - None: disables caching.
+
+local policy = require('apicast.policy')
+local _M = policy.new('Caching policy')
+
+local new = _M.new
+
+local function strict_handler(cache, cached_key, status_code, ttl)
+  if status_code == 200 then
+    ngx.log(ngx.INFO, 'apicast cache write key: ', cached_key,
+                      ', ttl: ', ttl, ' sub: ')
+    cache:set(cached_key, 200, ttl or 0)
+  else
+    ngx.log(ngx.NOTICE, 'apicast cache delete key: ', cached_key,
+                        ' cause status ', status_code)
+    cache:delete(cached_key)
+  end
+end
+
+local function resilient_handler(cache, cached_key, status_code, ttl)
+  if status_code and status_code < 500 then
+    ngx.log(ngx.INFO, 'apicast cache write key: ', cached_key,
+                      ' status: ', status_code, ', ttl: ', ttl)
+
+    cache:set(cached_key, status_code, ttl or 0)
+  end
+end
+
+local function disabled_cache_handler()
+  return function() end
+end
+
+local function handler(caching_type)
+  if caching_type == 'resilient' then
+    return resilient_handler
+  elseif caching_type == 'strict' then
+    return strict_handler
+  else
+    return disabled_cache_handler
+  end
+end
+
+local function init_config(config)
+  local res = {}
+
+  if not config.caching_type then
+    ngx.log(ngx.ERR, 'Caching type not specified. Cache disabled.')
+    res.caching_type = 'none'
+  elseif config.caching_type == 'resilient' or
+      config.caching_type == 'strict' or
+      config.caching_type == 'none' then
+    res.caching_type = config.caching_type
+  else
+    ngx.log(ngx.ERR, 'Invalid caching type. Cache disabled.')
+    res.caching_type = 'none'
+  end
+
+  return res
+end
+
+--- Initialize a Caching policy.
+-- @tparam[opt] table config
+-- @field caching_type Caching type (strict, resilient)
+function _M.new(config)
+  local self = new()
+  self.config = init_config(config or {})
+  return self
+end
+
+--- Access
+-- @tparam table context
+-- Overrides the cache_handler method of the proxy object included in the
+-- context. If no proxy object is received, it raises an error.
+function _M:access(context)
+  if not context.proxy then
+    ngx.log(ngx.ERR, 'Proxy not defined in the context')
+    return
+  end
+
+  context.proxy.cache_handler = handler(self.config.caching_type)
```

Comment:
What if this policy would just export the correct key in `:export()` method?
Then the APIcast policy could set the `cache_handler` to the proxy if it is in the context.

Then the "ownership"  of the proxy object would still be in the APIcast module. 
I'd be a bit worried that this is pretty deep access into internals of a module from some unrelated part of code. APIcast module initializes the Proxy object, so it could pass the cache handler as an argument for example.

Reply:
Definitely. I was thinking about the phases (`rewrite`, `access`, etc.) and forgot about `export()`.

Reply:
It turns out that `export()` is not working correctly for policies included in the local policy chain.

I suggest we address that problem on a separate PR. I've been investigating a bit and didn't find a straightforward solution.

To address your concern, in the rewrite phase of the cache policy, I included the cache_handler in the context. Later, the Apicast policy checks if it has been included, and if so, overrides the cache_handler method of the instance of Proxy that it owns.